### DISCUSSION TITLE
Suggest improvements or new features for CSSBattle

### DISCUSSION THREAD

**Discussion by kushsolitary** [cssbattle/forum#18] [Ideas]
Hi there,

We are having so much fun seeing everybody play and compete on CSSBattle. This battle has already been played by 1300+ players, which is the highest since our pilot battle. 🎊

We wanted to thank you and ask if you had any improvement or new features suggestions—anything which will make your experience with CSSBattle even better. Let us know here, and we'll start implementing the most upvoted ideas!

**Comment by iamvdo** [comment]
**Badges!** or rewards to unlock

Different ideas for these badges:

- full css solution (aka, no text or no html element)
- unicode or text-based (need to clarify what text-based means)
- less than 5 properties, less than 2 selectors, etc.
- no repeating colors more than once each
- special (advanced) properties: eg. blend modes, masks, etc.
- etc. etc.

It will add gamification, and also help users to "know" which tricks best solution is using.
Also, leaderboards could then be filtered using these badges

**Reply by chinchang** [reply]
wow! This is an excellent idea @iamvdo! It could solve many problems in one shot! CC @kushsolitary

**Reply by kushsolitary** [reply]
Thanks, @iamvdo. These are some very good suggestions.

**Reply by NaviTalib** [reply]
give me an example  of above lines. @iamvdo

**Comment by underdude** [comment]
While locking the battle leaderboard after battle ends makes sense and we point the glory to whom it belongs, I would like to have a toggle to show the current situation on battle level. As we've seen, even the oldest targets have been improved quite a lot since the battles have ended and the leaderboards would be a lot different now. Obviously especially the Battle 1, since the leaderboard is filled with SVG.

This might also give some motivation for new players to do some older battles and compare how they are doing within those targets. (Just a guess though, I've been here since the beginning :D )

**Reply by kushsolitary** [reply]
Very nice idea, we'll definitely consider it. Thank you!

**Comment by spalt08** [comment]
What about **infographics**? It might be interesting to create a stats/infographics page without any relation to specific targets. The data can be retrieved from the top 10/20 solutions from each target.

For example:
- Top 10 used CSS properties;
- Top 10 used selectors;
- Average amount of selectors in a solution;
- Average amount of HTML elements in a solution;
- etc.

**Reply by spalt08** [reply]
Personal stats might be interesting as well. You can create fancy profiles not only with rewards but with reach infographics (like gamers have for the Dota 2 for example).

It also could include average target solving time and total hours spent on the platform :)

**Reply by chinchang** [reply]
WoW! I like the idea of infographics page - showing stats from top solutions! Let's see what all interesting stuff we can get here. 
@spalt08 Do comment here if you get ideas on some more stats we can show.

**Reply by spalt08** [reply]
More ideas
- CSS selectors (quantity; distribution)
- CSS properties (quantity unique/total; distribution)
- Units distribution (q, %, px etc, bare numbers properties should be considered as px)
- CSS variables (avg. variables count)
- CSS gradients / background distribution (bare color, radial gradient, linear gradient etc.)
- Most popular HTML tags

**Comment by Esger** [comment]
I'm guilty too for copying and submitting the best solution on open targets, but this has the downside that there aren't shown many unique solutions. Moreover they're so advanced that beginners will have a hard time learning from them.
Maybe it would be nicer to show the top 10 scores while leaving out all duplicates, as it is now but there will always be 10 solutions to show, even if they are on a way lower position. Or maybe even show all unique solutions.

**Reply by Esger** [reply]
Why not show all unique solutions of the open targets? That would show way more of the different techniques to learn from.

**Reply by SelenIT** [reply]
That would require storing them in a separate DB structure, since the DB used doesn't have any analog to `SELECT DISTINCT` in SQL. It would probably worth doing so though!

**Reply by gunnarbittersmann** [reply]
FYI: [Evolution of top solutions of open targets](https://github.com/cssbattle/forum/discussions/31)

**Comment by iamvdo** [comment]
A preview tool, to verify solution as server will "see" it. Same as @alexzaworski's one, but right inside the UI

**Reply by spalt08** [reply]
It also can be a PRO feature because the implementation might cost a bit.

**Reply by chinchang** [reply]
@spalt08 Such a tool gives a direct advantage to the player. So this can't be PRO feature as it will be unfair to the non-PRO players. PRO features should only be luxury :)

**Reply by alexzaworski** [reply]
If it moves the needle at all I think there's clearly demand for this. It's a really insane advantage for people who are using strategies that rely on it, to an extent I'm not sure how people can compete without it.

In the last 30 days there's been 30k invocations on the function that drives previews 😄 which I'm sure is only a few people testing small changes a ton.

**Comment by iamvdo** [comment]
Hints when you have 99.9% match. It could be small circles where solution differs, on top of actual image, because sometimes it's hard to know why you won't get 100%, and what to improve.

**Reply by Esger** [reply]
Could be like the difference view in Alex's tool.

**Reply by underdude** [reply]
+1 for this, I usually add `mix-blend-mode: difference` manually to the div wrapping the preview iframe, it does the trick.

**Reply by iamvdo** [reply]
Oh nice. Thanks for this tip @underdude @Esger. Never used that, but seems like a great option. :)

**Reply by spalt08** [reply]
btw, what about custom plugins for pro version? Community can build advanced preview tools

**Reply by Beowolve** [reply]
> +1 for this, I usually add `mix-blend-mode: difference` manually to the div wrapping the preview iframe, it does the trick.

Nice trick, I will add this to my custom chrome plugin. Might release that in the near future. Has some other stuff as well.

**Reply by iamvdo** [reply]
That's the specific situation where it comes in handy to know where exactly. See #34.

**Reply by chinchang** [reply]
this has been now added to the play screen.

**Comment by Beowolve** [comment]
As someone relative new to the CSSBattle. I would love a chart visualizing my global ranking over time. Statistics on the profile page in general are a good idea as others already suggested. I really enjoyed climbing up the leaderboard, too addictive...

The Number of open targets for a running battle should be limited to just one during the battle I think. Now everyone just copies and pastes the top solution and the battle is limited to the other targets, there is really no point in optimizing the open targets other than having fun doing so. Keeping the sponsor's logo open is a good idea.

**Reply by SelenIT** [reply]
To me, open targets give a fun opportunity to quickly set a new record by finding some minor detail to improve (stray whitespace character, redundant selector, custom property with longer name used less times than one with shorter name, etc.) in the previous top solution. It's a good chance for new players to feel immediately rewarded for learning new optimisation tricks without spending much time to create shapes from scratch, and even for experienced players it's often helpful to skip the "boring" part of getting initial 100% match and start with the "fun" part of code golfing itself. But I'd find it super useful if each open target had a tool to observe its evolution, basically like manually updated pages [created](https://github.com/cssbattle/forum/discussions/31) by @gunnarbittersmann, but built-in and automated. I believe this is worth adding new DB structure for this, if necessary.

**Comment by M4N0JKUM4R** [comment]
**Monthly Leaderboard**

I was using TryHackMe for some time and it was too addictive with their point system which splits between Monthly and Yearly. It gave equal status to new comers and users who actively participate.

**Comment by ArpadGBondor** [comment]
I've solved them all, and uploaded the logic of my solutions to github. Didn't want to upload the compressed versions.
I tried staying within the "best solution +100%" character range, but halfway through it felt a bit boring and repetative. Just kept applying the same tricks, but I didn't feel that I'm improving, because I couldn't see other peoples solutions. The few top solutions, that's visible is more like CSS hacking, with unicode characters... couldn't really figured them out, and they didn't even look like the goal (at least in my browser), so they weren't really helpful for learning.

So, what I'd suggest is a way, to share your solutions, so others can see them, and leave comments, suggestions. A way to ask the top hackers about their solutions. Basically a way to learn from each other.

**Reply by aaron-cardenas** [reply]
If you have a 100% solution you should have access at least to the lower rated solutions. Of course with a different scoring system that could not be an issue. I like the way it works in code wars.

**Comment by aaron-cardenas** [comment]
A better SCORING system. Is not too late. It should enforce best practices and now do the opposite, and is tedious as hell.

**Comment by aaron-cardenas** [comment]
Slide & Compare whith option to hide de red line and horizontal/vertical slide

**Reply by Esger** [reply]
You can use shift for horizontal slide/compare

**Comment by Knusk** [comment]
I have a problem with getting a good score on code using mix-blend-mode. It 'seems' to be 100%, but I only get like 80% on it. This is true for [#53](https://cssbattle.dev/play/53) for example. So I suggest letting that kind of solution be valid when it 'looks' good, though the hex-code might be ever so slightly different colorwise. Other than that I find this very useful for learning, and lots of fun. Just a bit too addictive, maybe ... :)

**Reply by Esger** [reply]
I found a 100% solution with 168 chars and `mix-blend-mode:hard-light`.

**Reply by Knusk** [reply]
I think I made one other challenge work as well, and got 100%, so I don't know why it wouldn't work on this one. Did you get it to work with blend-mode on #53 also?

**Reply by gunnarbittersmann** [reply]
> I have a problem with getting a good score on code using mix-blend-mode. It 'seems' to be 100%, but I only get like 80% on it. This is true for [#53](https://cssbattle.dev/play/53) for example.

A mix-blend-mode-based solution had worked with the old backend. With the upgrade, it stopped working. But it’s still a 96% match, not only 80%. However, I’ve found a shorter solution not using mix-blend-mode.

**Reply by Esger** [reply]
I applied that on #53 only.

**Reply by gunnarbittersmann** [reply]
> I applied that on #53 only.

There’s 2 missed opportunities in one single sentence 😜:
- don’t use `mix-blend-mode` on a target where there are shorter solutions without
- use `mix-blend-mode` on a target where is leads to shorter solutions

**Reply by Knusk** [reply]
> > I have a problem with getting a good score on code using mix-blend-mode. It 'seems' to be 100%, but I only get like 80% on it. This is true for [#53](https://cssbattle.dev/play/53) for example.
> 
> A mix-blend-mode-based solution had worked with the old backend. With the upgrade, it stopped working. But it’s still a 96% match, not only 80%. However, I’ve found a shorter solution not using mix-blend-mode.

Yes, it was more like 89%, but still?

Here is my code that did not work.
`<div a></div><div b></div><div c><style>body{background:#19191A;margin:50 125}[a]{width:150;height:150;background:#4F77FF;border-radius:50%;mix-blend-mode:hard-light}[b]{margin-top:-75;width:75;height:125;background:#F9E492}`

**Reply by iamvdo** [reply]
I think it's a Chrome bug (the preview is in an `<iframe>` but not the server rendering)
Read: https://dev.to/bcalou/the-trouble-with-chrome-iframes-and-mix-blend-mode-4bi1

**Reply by Knusk** [reply]
Wow! Good catch! That explains why it didn't work for sure! I will try it in another browser asap ... :)

> I think it's a Chrome bug (the preview is in an `<iframe>` but not the server rendering)
> Read: https://dev.to/bcalou/the-trouble-with-chrome-iframes-and-mix-blend-mode-4bi1

**Reply by Knusk** [reply]
It did the trick! I had to change a few things; but now I could see what was wrong, and find out how to fix that. In Chrome it looked very much like 100%, but obviously was not. The most important part of this was that I was missing 'isolation: isolate' on the container; otherwise the body background was taken into consideration as well ...

**Comment by Esger** [comment]
When the battle is almost ending, the counter counts down to zero and then after it has ended it starts counting up again, there's no visual feedback that the battle has ended other then the counter going through zero.
![Schermafbeelding 2021-05-29 om 13 01 23](https://user-images.githubusercontent.com/70047/120068005-cd6fca80-c07e-11eb-9086-4c9189dc8c73.png)

**Reply by Esger** [reply]
This screenshot was taken 1 minute and 23 seconds after the battle ended.

**Reply by chinchang** [reply]
oh shoot! Fixing this...

**Reply by chinchang** [reply]
fixed now!

**Reply by Esger** [reply]
Some confetti or a checkered flag would be nice too :)

**Comment by hahagu** [comment]
An intended solution after the contest ends!
I sometimes really can't figure out how to do certain things in CSS, and as a user who uses CSS Battle as a means to practice and learn CSS, I think giving the solution that was intended by the host will help users of my kind very much!
I think it also goes without saying that the intended solution shouldn't be minified or super optimized beyond readability. :)

**Reply by Knusk** [reply]
I would very much appreciate that as well! The 100% solution that shows the 'best' way of solving such problems as the puzzle presents, would be very nice! Another thing might be to show how some of that could be minified, but that is not as important IMO, as the whole purpose of doing this is to gamify learning CSS.

**Reply by gunnarbittersmann** [reply]
You’re assuming the hosts have solutions to the targets they give us to solve? 🤣

**Reply by hahagu** [reply]
😂 I mean, they should've tested to see if it was possible at least once right..? lmao

**Reply by Knusk** [reply]
We'll see how it turns out with the CSSBattle _**Learn**_ part of this ... 😃

**Reply by Evavic44** [reply]
Lol.. 😄 So true... But that's a great idea. Though I'll suggest the perfect solution should only be available when you submit or timeout, so challengers can brainstorm on it first instead of just looking it up and copying it which will defeat it's purpose in the first place.

**Comment by Evavic44** [comment]
One Improvement I would love to see is adding IDE keyboard shortcuts (vs code in my case), can be very useful to improve performance and typing speed (Especially in the timed challenges). Maybe something to consider.

**Comment by Beowolve** [comment]
How about adding an "Activity Timeline", filtered by friends. So that you get informed if one of them managed to improve a target.

For PRO Members:
The current analytics page shows difference by character count, I would love to see a score difference.
Also, a more compact table would be more useful. Just a table with my score / character count and the top score / character count in each individual column.
Having everything listed side by side is not a very good choice for that. I would also like a CSV export function or even better an API to query the current statistics.
PRO needs more features for me to consider an extension of my subscription.

**Reply by chinchang** [reply]
Hey @Beowolve 👋🏼 We are working on some of the things mentioned here. Stay tuned for some updates.

**Comment by Esger** [comment]
It's cool we can use the VSCode editor as PRO badgers but the theme colors without are so much nicer and more readable. 
I'd really wish for the same theme-colors in the VSC editor.

**Comment by programus** [comment]
I hope CSS Battle can compress code automatically while scoring instead of let the coder do it, so that I can keep my code well formatted and easy to read. I always think the readability is as important as the functionality of a piece of code. 

Also, I think there could be two or more scoring system, of course which mean more ranks. One is the current one, the shortest code win; the other is a score with readability consideration, include well formatted, well naming points also. Of course, a badge system from @iamvdo is a great alternative of my idea. 

In the real industry world, functional code that well formatted and easy to read but may be not so efficient are usually much more welcome than extremely efficient codes but nobody (sometimes even the one who wrote it :-P ) understand it in some area are not so performance sensitive.

**Comment by programus** [comment]
I suggest a mini measure tool in the target view to let one easily know how many pixels the width, height, diameter, etc of an element, so we can concentrate on the CSS solutions not the measuring. 

Even a simple cursor's coordination related to the (0, 0) of the image would be much more helpful. 
For example:
![image](https://user-images.githubusercontent.com/956125/147431509-569f03fa-7bc0-468a-bd7d-90780200b5b1.png)

**Reply by Esger** [reply]
You can do that when you're on a Mac with Command Shift 4 (to make a screenshot from part of your screen) Just measure the area you want and press Escape to not make the screenshot or just release the mouse button if you don't care about screenshots on your desktop :)
Not sure if Windows has such a feature.

**Reply by programus** [reply]
> You can do that when you're on a Mac with Command Shift 4 (to make a screenshot from part of your screen) Just measure the area you want and press Escape to not make the screenshot or just release the mouse button if you don't care about screenshots on your desktop :)
> Not sure if Windows has such a feature.

Actually I am doing so when I play CSS battle on my Mac. When I play on my Windows laptop, I am using a tiny tool I developed many years ago. (Btw. You can download it here https://www.softpedia.com/get/Desktop-Enhancements/Other-Desktop-Enhancements/Programus-Screen-Ruler.shtml ) I developed it for my wife then, now it helps me also. But there’s a bug since Windows vista. 

So I consider both ways are not easy enough comparing add a coordination of cursor. I think it would not so difficult to implement. Maybe I can develop a chrome extension in the future for this if CSS Battle rejected this request and I am still playing it.

**Comment by gunnarbittersmann** [comment]
A **sort option** for “My solutions”: by date (as is currently), by length (characters), and by score (points). Maybe also by match percentage.
No need for ascending/descending order switch here; just newest/best first (i.e. date, score, percentage descending; length ascending).
Also no need for non-pro accounts now that the history is cut to five entries, so this might be a pro feature.

**Comment by snowiewdev** [comment]
It would be cool if the profile page could allow us to add LinkedIn profile/link. That would enable us to expand our professional network, and for the 'available for hire' tag to work better ;)
thank you :D

**Comment by M4N0JKUM4R** [comment]
I still don't see something like "Draft" so that it autosaves the current code even when something was incomplete, or am I missing something?

**Comment by APerricone** [comment]
I would like see all my scores and sort it from lowest to highest so I will work on worst

**Comment by SelenIT** [comment]
Some players express disappointment about old records that are not reachable anymore because of backend updates (e.g. 82-char solutions for [#30](https://cssbattle.dev/play/30)). Wouldn't it be useful to re-evaluate the top scores after each backend update, somewhat like it was done after disallowing the SVG solutions before the 2nd Battle? It can boost some players' interest in rechecking the already-played targets, exploring new ways to solve them, trying to restore the place in the leaderboard or, conversely, to win the new place that seemed unreachable before the update.

**Comment by Esger** [comment]
It would be very convenient to have '< previous' and 'next >' buttons around the #target dropdown, so one can browse through all the targets and battles directly. These buttons decrease or increase the battle #. Or shortcuts to do that.

**Comment by Beowolve** [comment]
I do have an Idea that could help newer players and also make the game in general more interresting.

There are now more then 100 targets and it keeps getting more every battle. So new players might get overwhelmed by the big number of targets.

I suggest making older battles public every time a new closed target gets added. So that we have 100 closed targets all the time and not more.

* This would show more tricks and methods in the long run
* New players don't have to do all the targets on there own if they don't want to
* Leaderboards would be more dynamic
* There is more to learn for everyone
* Old records will be broken for sure!
* Also more streaming potential 😎 

Only downside with that, we topplayers can't keep our nice records. But I am ok with that, there are new records to be made anyway 😃.

**Reply by Esger** [reply]
In the long run that will converge the scores of all active players towards the top. Not sure if that is wanted.

**Reply by Beowolve** [reply]
Why would that be? Every new closed target will also  create a new leader score there (expect for current battle ;D)
The only thing that would close the gap a little, are some  tricks that might get released. But I don't think that will make a huge difference, most tricks are already known. It's mostly a matter of using them right and being creative.

**Reply by chinchang** [reply]
At some point, in the long run, we would have 1000 open targets and 100 closed targets. So overall score would have more and more open targets as its contribution over time. That means same score on open targets, hence very similar overall score, hence the convergence.

**Reply by Esger** [reply]
Another way to achieve all of your goals is opening al sub top 100 solutions. It will reduce the convergence a bit as well.

**Comment by diamdesign** [comment]
- Time challenges showing elapsed time while working, also showing elapsed time on the leaderboard, also adding it to the score, making the ones who make it faster have a higher score.
- Improve protection against the "hacker" bug, where users use "fail" solutions to imitate the image in only one faulty browser. Meaning you got the message "may not work in all browsers" then that's not a 100% match is it? Opening a solution and seeing it's not 100% match, yet your system has put them on the no1 leaderboard, makes your leaderboard faulty.
- Add animation challenges, to create your own html/css challenges. So I can paste HTML and CSS, being able to do any type of CSS animations. And with that, your system should show the html/css animation, maybe as a video/gif, so that people can't check the code...
- If there is a way; give more points if the solution actually is a 100% match in more than one browser.

**Reply by APerricone** [reply]
> * Improve protection against the "hacker" bug, where users use "fail" solutions to imitate the image in only one faulty browser. Meaning you got the message "may not work in all browsers" then that's not a 100% match is it? Opening a solution and seeing it's not 100% match, yet your system has put them on the no1 leaderboard, makes your leaderboard faulty.

I think you are a little confused about it, the system do a rendering on the server of your code, that system can be different from your so the result can be very different, There is no "hacker bug". To get a result equals (or very similar) to server rendering you can use the [cssbattle previewer](https://tc70f3.csb.app/)

**Comment by AlRey-Dev** [comment]
[1] Determine which battles are "untouched" by select friends and/or other players. This is to support/complement the idea below...

[2] The ability to generate a "competitive/performance report/statistics" against select friends and/or players (for example: when organizing a random CSSBattle contest with other players (e.g. work colleagues) that you haven't added yet as a friend).

inputs:
     - battle numbers
     - player tags
     - contest duration (optional)
output: 
     contest-based context:
          - fastest to solve (i.e. 100%) all "battle numbers" input
          - rankings based on number of solved battles + started but <100%
     player-based context
          - remaining <100% battles
          - remaining battles that are not started
     contest organizer-based context:
          - registered players (all players in the "player tags" input) vs attending players (those with at least one non-zero submission) vs no-shows (those with zero submissions)

**Comment by joewinbpaul** [comment]
how about the thing I suggested in #336?
